### PR TITLE
Updated ThrottledBackgroundMessageProcessor to dynamically create task workers

### LINKED ISSUE #541: Thread pool exhaustion on version 11 of `Mindscape.Raygun4Net.AspNetCore`
**Describe the bug**
Since upgrading to version 11 we have been seeing our instances suffer from what we believe to be thread pool exhaustion. The problem is apparent on versions `11.0.2` and `11.0.3` of `Mindscape.Raygun4Net.AspNetCore` but does not appear when we downgraded back to our previous version (`9.0.4`).

This is manifesting in slower HTTP responses from one of our services, which we noticed when the service started failing liveness/ready probes. This is a fairly simple service that has a few endpoints and calls a few external APIs when hit, but generally doesn't do that much business logic.

On further inspection, we noticed that when instances are started they seem to be in a "good" state (no slowdowns, no liveness probe failures), but they can be flipped into a "bad" state (some slowdowns, persistent periodic liveness probe failures). The only other signifier of this bad state is that the thread count metrics were higher for these instances and so was the threadpool queue length. Furthermore, the timing of when instances got shifted into this bad state was correlated with when those instances sent their first raygun error.

We ran `dotnet-stack` on an affected instance, and the majority of the threads in progress (8 in total) had stacks like the following:
```
Thread (0x1DA1):
  [Native Frames]
  System.Private.CoreLib!System.Threading.SemaphoreSlim.WaitUntilCountOrTimeout(int32,unsigned int32,value class System.Threading.CancellationToken)
  System.Private.CoreLib!System.Threading.SemaphoreSlim.Wait(int32,value class System.Threading.CancellationToken)
  System.Collections.Concurrent!System.Collections.Concurrent.BlockingCollection`1[System.__Canon].TryTakeWithNoTimeValidation(!0&,int32,value class System.Threading.CancellationToken,class System.Threading.CancellationTokenSource)
  System.Collections.Concurrent!System.Collections.Concurrent.BlockingCollection`1+<GetConsumingEnumerable>d__68[System.__Canon].MoveNext()
  Mindscape.Raygun4Net.NetCore.Common!Mindscape.Raygun4Net.ThrottledBackgroundMessageProcessor+<RaygunMessageWorker>d__14.MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.ThrottledBackgroundMessageProcessor+<RaygunMessageWorker>d__14].ExecutionContextCallback(class System.Object)
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.ThrottledBackgroundMessageProcessor+<RaygunMessageWorker>d__14].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.ThrottledBackgroundMessageProcessor+<RaygunMessageWorker>d__14].MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.Threading.Tasks.VoidTaskResult].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
  Mindscape.Raygun4Net.NetCore.Common!Mindscape.Raygun4Net.RaygunClientBase+<Send>d__57.MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<Send>d__57].ExecutionContextCallback(class System.Object)
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<Send>d__57].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<Send>d__57].MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.Threading.Tasks.VoidTaskResult].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
  Mindscape.Raygun4Net.NetCore.Common!Mindscape.Raygun4Net.RaygunClientBase+<SendPayloadAsync>d__59.MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<SendPayloadAsync>d__59].ExecutionContextCallback(class System.Object)
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<SendPayloadAsync>d__59].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,Mindscape.Raygun4Net.RaygunClientBase+<SendPayloadAsync>d__59].MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Threading.Tasks.Task`1[System.__Canon].TrySetResult(!0)
  System.Net.Http!System.Net.Http.HttpClient+<<SendAsync>g__Core|83_0>d.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.__Canon,System.Net.Http.HttpClient+<<SendAsync>g__Core|83_0>d].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.__Canon].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Net.Http!System.Net.Http.RedirectHandler+<SendAsync>d__4.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.__Canon,System.Net.Http.RedirectHandler+<SendAsync>d__4].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Threading.Tasks.Task`1[System.__Canon].TrySetResult(!0)
  System.Net.Http!System.Net.Http.DiagnosticsHandler+<SendAsyncCore>d__8.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.__Canon,System.Net.Http.DiagnosticsHandler+<SendAsyncCore>d__8].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Threading.Tasks.Task`1[System.__Canon].TrySetResult(!0)
  System.Net.Http!System.Net.Http.HttpConnectionPool+<SendWithVersionDetectionAndRetryAsync>d__84.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.__Canon,System.Net.Http.HttpConnectionPool+<SendWithVersionDetectionAndRetryAsync>d__84].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.__Canon].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Net.Http!System.Net.Http.HttpConnection+<SendAsyncCore>d__64.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.__Canon,System.Net.Http.HttpConnection+<SendAsyncCore>d__64].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Threading.Tasks.Task`1[System.Threading.Tasks.VoidTaskResult].TrySetResult(!0)
  System.Net.Http!System.Net.Http.HttpConnection+<InitialFillAsync>d__101.MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult,System.Net.Http.HttpConnection+<InitialFillAsync>d__101].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.Int32].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Net.Security!System.Net.Security.SslStream+<ReadAsyncInternal>d__188`1[System.Net.Security.AsyncReadWriteAdapter].MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Int32,System.Net.Security.SslStream+<ReadAsyncInternal>d__188`1[System.Net.Security.AsyncReadWriteAdapter]].MoveNext(class System.Threading.Thread)
  System.Private.CoreLib!System.Runtime.CompilerServices.TaskAwaiter+<>c.<OutputWaitEtwEvents>b__12_0(class System.Action,class System.Threading.Tasks.Task)
  System.Private.CoreLib!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(class System.Action,bool)
  System.Private.CoreLib!System.Threading.Tasks.Task.RunContinuations(class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[System.Int32].SetExistingTaskResult(class System.Threading.Tasks.Task`1<!0>,!0)
  System.Net.Security!System.Net.Security.SslStream+<EnsureFullTlsFrameAsync>d__186`1[System.Net.Security.AsyncReadWriteAdapter].MoveNext()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Int32,System.Net.Security.SslStream+<EnsureFullTlsFrameAsync>d__186`1[System.Net.Security.AsyncReadWriteAdapter]].MoveNext(class System.Threading.Thread)
  System.Net.Sockets!System.Net.Sockets.Socket+AwaitableSocketAsyncEventArgs.InvokeContinuation(class System.Action`1<class System.Object>,class System.Object,bool,bool)
  System.Net.Sockets!System.Net.Sockets.SocketAsyncEngine.System.Threading.IThreadPoolWorkItem.Execute()
  System.Private.CoreLib!System.Threading.ThreadPoolWorkQueue.Dispatch()
  System.Private.CoreLib!System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()
```

Running `dotnet-trace` on the instance using the `cpu-sampler` profile indicated that there were 7 threads spending most of their time in Raygun code, presumably waiting on something.

At this point we downgraded the package and the problems have not reoccurred since.

We don't have any customisations to the base settings, we're initializing like this:

```
        services.AddRaygun(raygunSettings =>
        {
            raygunSettings.ApiKey = m_Configuration.GetValue<string>("RAYGUN_API_KEY");
        });
```

### PR DESCRIPTION
**Pull request by phillip-haydon**
Should resolve #541 where up to 8 tasks workers can sit idle doing nothing when there's no messages to be sent. This change will create the task workers on demand based on the queue length and dispose of them if idle.

### CODE REVIEW

**Review by QuantumNightmare — DISMISSED** [review]
Took a bit to follow this but looks like it'll do the trick. Cancelling a worker causes it to complete its current task before disposing right? Make sure to do a test of spamming a fixed number of exception reports and confirming that the correct number of exceptions show up in Raygun. Maybe force exceptions to occur at certain points of the worker processing and check that it behaves as expected as well.

**Review by MattByers — COMMENTED** [review]
I can't see any immediately obvious issues. Same as Jase mentioned though, I am curious about the behaviour that happens, when a task that is currently doing work is cancelled. Does it definitely complete it's current message?

Another thought. This class looks to be duplicated at [Mindscape.Raygun4Net4.ThrottledBackgroundMessageProcessor](https://github.com/MindscapeHQ/raygun4net/blob/master/Mindscape.Raygun4Net4/ThrottledBackgroundMessageProcessor.cs). Is it worth carrying these changes over there, so it helps with the idle tasks for .net framework.

**Comment by phillip-haydon** [comment]
> Took a bit to follow this but looks like it'll do the trick. Cancelling a worker causes it to complete its current task before disposing right? Make sure to do a test of spamming a fixed number of exception reports and confirming that the correct number of exceptions show up in Raygun. Maybe force exceptions to occur at certain points of the worker processing and check that it behaves as expected as well.

I ran a benchmark with 100k messages and found a race condition during the draining which caused ~1-5 messages to get stuck and not processed. This happened because when the AdjustWorkers is run, it grabs a count and once a count is obtained, but before the adjustment occurs, a message could come in and be queued, but we only allow 1 thread to enter the method at a time.

Solution was to just double-check the queue size before exiting the AjustWorkers and the issue goes away. 

Besides this it appears to work well. In all cases the queue gets processed quicker than the old implementation, but somewhere around ~1k messages in the queue the memory usage grows a bit more than the old implementation, I think this is due to the Concurrent Queue resizing internally. But at 100k messages the difference looks to be ~5mb difference (45mb vs 50mb)

This gets cleaned up tho when the queue is processed. 

In the old implementation we used a BlockingCollection which has a capacity argument, the ConcurrentQueue doesn't have this, but I have a check in the Queue with a drain percentage so if we hit the capacity it will drain 10% before accepting anymore messages. Give it some breathing room.

**Comment by phillip-haydon** [comment]
> Another thought. This class looks to be duplicated at [Mindscape.Raygun4Net4.ThrottledBackgroundMessageProcessor](https://github.com/MindscapeHQ/raygun4net/blob/master/Mindscape.Raygun4Net4/ThrottledBackgroundMessageProcessor.cs). Is it worth carrying these changes over there, so it helps with the idle tasks for .net framework.

I only want to deal with the .NET Core version for now, may consider back-porting it to .NET Framework in the future.

File: Mindscape.Raygun4Net.NetCore.Tests/ThrottledBackgroundMessageProcessorTests.cs [hunk 1]
```diff
@@ -148,5 +154,71 @@ public void ThrottledBackgroundMessageProcessor_CancellationRequested_IsCaughtAn
...
+          Console.WriteLine(i);
+        }
+
+        resetEventSlim.Wait(TimeSpan.FromSeconds(10));
+
+        cut.Dispose();
+      }
+
+      Assert.That(secondMessageWasProcessed, Is.True);
+      //Assert.That(secondMessageWasProcessed, Is.True);
```
Comment (ProRedCat) [inline Mindscape.Raygun4Net.NetCore.Tests/ThrottledBackgroundMessageProcessorTests.cs:221]: Remove comment

File: Mindscape.Raygun4Net.NetCore.Tests/ThrottledBackgroundMessageProcessorTests.cs [hunk 2]
```diff
@@ -148,5 +154,71 @@ public void ThrottledBackgroundMessageProcessor_CancellationRequested_IsCaughtAn
...
+
+        cut.Enqueue(new RaygunMessage());
+
+        resetEventSlim.Wait(TimeSpan.FromSeconds(5));
+
+        cut.Dispose();
+      }
+
+      Assert.That(secondMessageWasProcessed, Is.True);
+      //Assert.That(secondMessageWasProcessed, Is.True);
```
Comment (ProRedCat) [inline Mindscape.Raygun4Net.NetCore.Tests/ThrottledBackgroundMessageProcessorTests.cs:183]: Comment

**Review by ProRedCat — APPROVED** [review]
Lgtm, this appears to have fixed some issues in Raygun4Maui too

**Review by ProRedCat — APPROVED** [review]

**Pull request merged**